src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.cs

```diff
@@ -78,39 +78,40 @@ internal static int CompareStringIgnoreCaseNonAscii(ref char strA, int lengthA,
             return OrdinalCasing.CompareStringIgnoreCase(ref strA, lengthA, ref strB, lengthB);
         }
 
-        private static bool EqualsIgnoreCase_Vector128(ref char charA, ref char charB, int length)
+        private static bool EqualsIgnoreCase_Vector<TVector>(ref char charA, ref char charB, int length)
+            where TVector : struct, ISimdVector<TVector, ushort>
         {
-            Debug.Assert(length >= Vector128<ushort>.Count);
-            Debug.Assert(Vector128.IsHardwareAccelerated);
+            Debug.Assert(length >= TVector.Count);
+            Debug.Assert(TVector.IsHardwareAccelerated);
 
             nuint lengthU = (nuint)length;
-            nuint lengthToExamine = lengthU - (nuint)Vector128<ushort>.Count;
+            nuint lengthToExamine = lengthU - (nuint)TVector.Count;
             nuint i = 0;
-            Vector128<ushort> vec1;
-            Vector128<ushort> vec2;
+            TVector vec1;
+            TVector vec2;
             do
             {
-                vec1 = Vector128.LoadUnsafe(ref charA, i);
-                vec2 = Vector128.LoadUnsafe(ref charB, i);
+                vec1 = TVector.LoadUnsafe(ref Unsafe.As<char, ushort>(ref charA), i);
+                vec2 = TVector.LoadUnsafe(ref Unsafe.As<char, ushort>(ref charB), i);
 
-                if (!Utf16Utility.AllCharsInVector128AreAscii(vec1 | vec2))
+                if (!Utf16Utility.AllCharsInVectorAreAscii(vec1 | vec2))
                 {
                     goto NON_ASCII;
                 }
 
-                if (!Utf16Utility.Vector128OrdinalIgnoreCaseAscii(vec1, vec2))
+                if (!Utf16Utility.VectorOrdinalIgnoreCaseAscii(vec1, vec2))
                 {
                     return false;
                 }
 
-                i += (nuint)Vector128<ushort>.Count;
+                i += (nuint)TVector.Count;
```

Comment:
This might noticeably regress comparisons for `length < TVector.Count` when AVX512 fully lights up, and a lot of string comparisons might fall under such a scenario (< 32 chars).

Naturally, this poses a challenge of expressing TVector vectorized core, with TVector / 2 (e.g. AVX512 -> AVX2) fallback to ensure ~10-30 character strings are compared quickly.

I'm not sure what would the good catch-all solution for such problem look like given how most vectorized loops are currently implemented in CoreLib, but it might become more of a problem in the future, so I wanted to highlight this.

Reply:
The implementation can be updated to handle trailing elements via SIMD too

Reply:
Data alignment is worth doing but it's sadly quite verbose to do. It's especially beneficial to do for string opts because chars in string objects are only 4-byte aligned.

Reply:
> Data alignment is worth doing but it's sadly quite verbose to do. It's especially beneficial to do for string opts because chars in string objects are only 4-byte aligned.

Would we pin to guarantee we're keeping alignment for the whole loop or would we just align at the start and hope the GC doesn't move stuff if we'd be adding that to SIMD loops in .Net?

Reply:
> > Data alignment is worth doing but it's sadly quite verbose to do. It's especially beneficial to do for string opts because chars in string objects are only 4-byte aligned.
> 
> Would we pin to guarantee we're keeping alignment for the whole loop or would we just align at the start and hope the GC doesn't move stuff if we'd be adding that to SIMD loops in .Net?

I guess pinning is the only option